Release/src/http/common/http_helpers.cpp

```diff
@@ -427,5 +431,310 @@ bool validate_method(const utility::string_t& method)
+
+            return input_offset;
+        }
+
+        bool is_complete() const
+        {
+            return state() == Z_STREAM_END;
+        }
+
+        bool has_error() const
+        {
+            return !is_complete() && state() != Z_OK;
+        }
+
+        int state() const
+        {
+            return m_zLibState;
+        }
+
+        void set_state(int state)
+        {
+            m_zLibState = state;
+        }
+
+        compression_algorithm algorithm() const
+        {
+            return m_alg;
+        }
+
+        z_stream& stream()
+        {
+            return m_zLibStream;
+        }
+
+        int to_zlib_alg(compression_algorithm alg)
+        {
+            return static_cast<int>(alg);
+        }
+
+    private:
+        const compression_algorithm m_alg;
+
+        std::atomic<int> m_zLibState{ Z_OK };
+        z_stream m_zLibStream;
+    };
+
+    class stream_decompressor::stream_decompressor_impl : public compression_base_impl
+    {
+    public:
+        stream_decompressor_impl(compression_algorithm alg) : compression_base_impl(alg)
+        {
+            set_state(inflateInit2(&stream(), to_zlib_alg(alg)));
+        }
+
+        ~stream_decompressor_impl()
+        {
+            inflateEnd(&stream());
+        }
+
+        data_buffer decompress(const uint8_t* input, size_t input_size)
+        {
+            if (input == nullptr || input_size == 0)
+            {
+                set_state(Z_BUF_ERROR);
+                return data_buffer();
+            }
+
+            // Need to guard against attempting to decompress when we're already finished or encountered an error!
+            if (is_complete() || has_error())
+            {
+                set_state(Z_STREAM_ERROR);
+                return data_buffer();
+            }
+
+            const size_t BUFFER_SIZE = 1024;
+            unsigned char temp_buffer[BUFFER_SIZE];
+
+            data_buffer output;
+            output.reserve(input_size * 3);
+
+            size_t input_offset{ 0 };
+
+            while (state() == Z_OK && input_offset < input_size)
+            {
+                auto total_out_before = stream().total_out;
+
+                auto available_input = input_size - input_offset;
+                stream().next_in = const_cast<uint8_t*>(&input[input_offset]);
+                stream().avail_in = static_cast<int>(available_input);
+                stream().next_out = temp_buffer;
+                stream().avail_out = BUFFER_SIZE;
+
+                set_state(inflate(&stream(), Z_PARTIAL_FLUSH));
+
+                if (has_error())
+                {
+                    return data_buffer();
+                }
+
+                input_offset = read_output(input_offset, available_input, total_out_before, temp_buffer, output);
+            }
+
+            return output;
+        }
+    };
+
+    class stream_compressor::stream_compressor_impl : public compression_base_impl
```

Comment:
This isnt actually used in production code (yet). Sending compressed content in a request isn't strictly standardised. There was a request in the github issue to support this however. It is also needed for one of the tests below.